include/watchdog.h

```diff
@@ -1,116 +1,219 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
+ * Copyright (c) 2017 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
+/**
+ * @brief Watchdog Interface
+ * @defgroup watchdog_interface Watchdog Interface
+ * @ingroup timer_counter_interfaces
+ * @{
+ */
+
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
+/**
+ * @brief Pause watchdog timer when CPU is in sleep state.
+ */
+#define WDT_OPT_PAUSE_IN_SLEEP	BIT(0)
 
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
-};
+/**
+ * @brief Pause watchdog timer when CPU is halted by the debugger.
+ */
+#define WDT_OPT_PAUSE_HALTED_BY_DBG	BIT(1)
 
 /**
- * WDT clock cycles for timeout type.
- */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
-};
+ * @brief Run callback function on watchdog timeout. If any of reset flags is
+ *	  set this callback is called before reset.
+ */
+#define WDT_FLAG_RUN_CALLBACK	BIT(0)
+
+/**
+ * @brief Watchdog reset behavior.
+ *
+ * If none of the reset flags is specified there will be no reset after timeout.
+ */
+/** CPU core reset */
+#define WDT_FLAG_RESET_CPU	BIT(1)
+/** CPU core and peripherals reset */
+#define WDT_FLAG_RESET_CPU_PER	BIT(2)
+/** Global system reset */
+#define WDT_FLAG_RESET_SOC	(BIT(2) | BIT(1))
```

Comment:
I do not know how to make clean group doxygen comment for these macros. I have tried different ways from doxygen manual however none on them worked with our auto generated html documentation.

Reply:
OK. I found a few issues here that will take a few commits to make work.  The way we were including the auto-generated docs used a :content-only: directive that on the surface looks OK, but has the side-effect of causing all the doxygen headings to be removed and the API material listed in alphabetic order (instead of separating the functions from the typedefs, for example).  Removing the :content-only: fixed that, but resulted in some extra headings and layout issues (that required a CSS tweak).  When all this gets merged, you can write this section like this to make it come out as a "group doxygen comment" (note the addition of the @name and the start/stop group markers @{ and @}:

     /**
      * @name Watchdog Reset Behavior
      * If none of the reset flags is specified there will be no reset after timeout.
      * @{
      */
     /** CPU core reset */
     #define WDT_FLAG_RESET_CPU>-----BIT(1)
     /** CPU core and peripherals reset */  
     #define WDT_FLAG_RESET_CPU_PER>-BIT(2)
     /** Global system reset */
     #define WDT_FLAG_RESET_SOC>-----(BIT(2) | BIT(1))
     /** @} */

Reply:
Ready to use this documentation grouping syntax now.